### PR TITLE
Choropleth Counties

### PR DESCRIPTION
**Pull request by Kully**
Re: https://github.com/plotly/plotly.py/issues/897

### CODE REVIEW

File: plotly/tests/test_optional/test_figure_factory.py [hunk 1]
```diff
@@ -2718,3 +2719,17 @@ def test_full_bullet(self):
...
+    def test_fips_values_same_len(self):
+        pattern = 'fips and values must be the same length'
+        self.assertRaisesRegexp(
+            PlotlyError, pattern, ff.create_choropleth, [1], [1, 2]
+        )
+
+
+
+
+
```
Comment (cldougl) [inline plotly/tests/test_optional/test_figure_factory.py]: there are a lot of extra newlines added at the bottom of this file

File: plotly/version.py [hunk 2]
```diff
@@ -1 +1 @@
-__version__ = '2.3.0'
+__version__ = '2.3.1'
```
Comment (cldougl) [inline plotly/version.py]: I believe this should be 2.4.0 https://semver.org/

File: plotly/tests/test_optional/test_figure_factory.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 
 import numpy as np
 from scipy.spatial import Delaunay
+import os
```
Comment (cldougl) [inline plotly/tests/test_optional/test_figure_factory.py]: did `os` need to be imported for the added choropleth test? It doesn't look like it below.

File: plotly/figure_factory/_county_choropleth.py [hunk 4]
```diff
@@ -0,0 +1,903 @@
...
+    if height / width > (1 / asp):
+        new_width = asp * height
+        fig['layout']['xaxis']['range'][0] = center[0] - new_width * 0.5
+        fig['layout']['xaxis']['range'][1] = center[0] + new_width * 0.5
+    else:
+        new_height = (1 / asp) * width
+        fig['layout']['yaxis']['range'][0] = center[1] - new_height * 0.5
+        fig['layout']['yaxis']['range'][1] = center[1] + new_height * 0.5
+
+    return fig
```
Comment (cldougl) [inline plotly/figure_factory/_county_choropleth.py]: newline at the end of file recommended

**Comment by jackparmer** [comment]
Do we need this xml file?
https://github.com/plotly/plotly.py/pull/929/files#diff-b69057916476b49938dab4751dc2413e
Are there any other files we should remove that we don't need?

**Comment by jackparmer** [comment]
Just to confirm, the current version of this FigureFactory in this PR generates these plots?
https://plot.ly/~AdamKulidjian/6680

File: plotly/figure_factory/_county_choropleth.py [hunk 5]
```diff
@@ -557,7 +557,7 @@ def create_choropleth(fips, values, scope=['usa'], binning_endpoints=None,
     """
     # ensure optional modules imported
     if not gp or not shapefile or not shapely:
-        raise ImportError("geopandas, shapefile and shapely must be "
+        raise ImportError("geopandas, pyshp and shapely must be "
                           "installed for this figure factory")
```
Comment (jackparmer) [inline plotly/figure_factory/_county_choropleth.py]: Should we add the pip statements here? For example, "Try pip install geopandas and pip install pyshp in your terminal."

**Comment by jackparmer** [comment]
There's a mismatch between this notebook and the plots. I think you updated the plots, but didn't upload the new notebook: https://plot.ly/~AdamKulidjian/6680 Mind uploading the latest?

**Comment by jackparmer** [comment]
💃 After considering the comments above. Great work on this one. By far the easiest way to make interactive choropleth maps in Python to the best of my knowledge. 🥂 

image

File: plotly/figure_factory/_county_choropleth.py (hunk 5 above)
Comment (Kully) [reply plotly/figure_factory/_county_choropleth.py]: good idea. we can add it to the docs as well to remind in two places

Comment (Kully) [reply plotly/figure_factory/_county_choropleth.py]: > Just to confirm

confirmed

**Comment by Kully** [comment]
> Mind uploading the latest?

Here's the latest notebook: https://plot.ly/~AdamKulidjian/6681

**Pull request merged**